File: tests/test_color_management_effects.py
"""Color management effects class test harness."""

import unittest
import opentimelineio as otio
import opentimelineio.test_utils as otio_test_utils

class VideoBrightnessTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        brightness = otio.schema.VideoBrightness(
            name="BrightIt",
            brightness=50,
            metadata={"foo": "bar"}
        )
        self.assertEqual(brightness.brightness, 50)
        self.assertEqual(brightness.name, "BrightIt")
        self.assertEqual(brightness.metadata, {"foo": "bar"})

    def test_eq(self):
        b1 = otio.schema.VideoBrightness(name="BrightIt", brightness=50, metadata={"foo": "bar"})
        b2 = otio.schema.VideoBrightness(name="BrightIt", brightness=50, metadata={"foo": "bar"})
        self.assertIsOTIOEquivalentTo(b1, b2)

    def test_serialize(self):
        brightness = otio.schema.VideoBrightness(name="BrightIt", brightness=50, metadata={"foo": "bar"})
        encoded = otio.adapters.otio_json.write_to_string(brightness)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(brightness, decoded)

    def test_setters(self):
        brightness = otio.schema.VideoBrightness(name="BrightIt", brightness=50, metadata={"foo": "bar"})
        self.assertEqual(brightness.brightness, 50)
        brightness.brightness = 100
        self.assertEqual(brightness.brightness, 100)

class VideoContrastTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        contrast = otio.schema.VideoContrast(
            name="ContrastIt",
            contrast=20,
            metadata={"foo": "bar"}
        )
        self.assertEqual(contrast.contrast, 20)
        self.assertEqual(contrast.name, "ContrastIt")
        self.assertEqual(contrast.metadata, {"foo": "bar"})

    def test_eq(self):
        c1 = otio.schema.VideoContrast(name="ContrastIt", contrast=20, metadata={"foo": "bar"})
        c2 = otio.schema.VideoContrast(name="ContrastIt", contrast=20, metadata={"foo": "bar"})
        self.assertIsOTIOEquivalentTo(c1, c2)

    def test_serialize(self):
        contrast = otio.schema.VideoContrast(name="ContrastIt", contrast=20, metadata={"foo": "bar"})
        encoded = otio.adapters.otio_json.write_to_string(contrast)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(contrast, decoded)

    def test_setters(self):
        contrast = otio.schema.VideoContrast(name="ContrastIt", contrast=20, metadata={"foo": "bar"})
        self.assertEqual(contrast.contrast, 20)
        contrast.contrast = 40
        self.assertEqual(contrast.contrast, 40)

class VideoSaturationTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        saturation = otio.schema.VideoSaturation(
            name="SaturateIt",
            saturation=70,
            metadata={"foo": "bar"}
        )
        self.assertEqual(saturation.saturation, 70)
        self.assertEqual(saturation.name, "SaturateIt")
        self.assertEqual(saturation.metadata, {"foo": "bar"})

    def test_eq(self):
        s1 = otio.schema.VideoSaturation(name="SaturateIt", saturation=70, metadata={"foo": "bar"})
        s2 = otio.schema.VideoSaturation(name="SaturateIt", saturation=70, metadata={"foo": "bar"})
        self.assertIsOTIOEquivalentTo(s1, s2)

    def test_serialize(self):
        saturation = otio.schema.VideoSaturation(name="SaturateIt", saturation=70, metadata={"foo": "bar"})
        encoded = otio.adapters.otio_json.write_to_string(saturation)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(saturation, decoded)

    def test_setters(self):
        saturation = otio.schema.VideoSaturation(name="SaturateIt", saturation=70, metadata={"foo": "bar"})
        self.assertEqual(saturation.saturation, 70)
        saturation.saturation = 100
        self.assertEqual(saturation.saturation, 100)

class VideoLightnessTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        lightness = otio.schema.VideoLightness(
            name="LightIt",
            lightness=10,
            metadata={"foo": "bar"}
        )
        self.assertEqual(lightness.lightness, 10)
        self.assertEqual(lightness.name, "LightIt")
        self.assertEqual(lightness.metadata, {"foo": "bar"})

    def test_eq(self):
        l1 = otio.schema.VideoLightness(name="LightIt", lightness=10, metadata={"foo": "bar"})
        l2 = otio.schema.VideoLightness(name="LightIt", lightness=10, metadata={"foo": "bar"})
        self.assertIsOTIOEquivalentTo(l1, l2)

    def test_serialize(self):
        lightness = otio.schema.VideoLightness(name="LightIt", lightness=10, metadata={"foo": "bar"})
        encoded = otio.adapters.otio_json.write_to_string(lightness)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(lightness, decoded)

    def test_setters(self):
        lightness = otio.schema.VideoLightness(name="LightIt", lightness=10, metadata={"foo": "bar"})
        self.assertEqual(lightness.lightness, 10)
        lightness.lightness = 20
        self.assertEqual(lightness.lightness, 20)

class VideoColorTemperatureTests(unittest.TestCase, otio_test_utils.OTIOAssertions):
    def test_constructor(self):
        temp = otio.schema.VideoColorTemperature(
            name="TempIt",
            temperature=6500,
            metadata={"foo": "bar"}
        )
        self.assertEqual(temp.temperature, 6500)
        self.assertEqual(temp.name, "TempIt")
        self.assertEqual(temp.metadata, {"foo": "bar"})

    def test_eq(self):
        t1 = otio.schema.VideoColorTemperature(name="TempIt", temperature=6500, metadata={"foo": "bar"})
        t2 = otio.schema.VideoColorTemperature(name="TempIt", temperature=6500, metadata={"foo": "bar"})
        self.assertIsOTIOEquivalentTo(t1, t2)

    def test_serialize(self):
        temp = otio.schema.VideoColorTemperature(name="TempIt", temperature=6500, metadata={"foo": "bar"})
        encoded = otio.adapters.otio_json.write_to_string(temp)
        decoded = otio.adapters.otio_json.read_from_string(encoded)
        self.assertIsOTIOEquivalentTo(temp, decoded)

    def test_setters(self):
        temp = otio.schema.VideoColorTemperature(name="TempIt", temperature=6500, metadata={"foo": "bar"})
        self.assertEqual(temp.temperature, 6500)
        temp.temperature = 7000
        self.assertEqual(temp.temperature, 7000)

File: tests/test_color_management_effects.cpp
#include "utils.h"

#include <opentimelineio/clip.h>
#include <opentimelineio/deserialization.h>
#include <opentimelineio/externalReference.h>
#include <opentimelineio/colorManagementEffects.h>

namespace otime = opentime::OPENTIME_VERSION;
namespace otio  = opentimelineio::OPENTIMELINEIO_VERSION;

int
main(int argc, char** argv)
{
    Tests tests;
    tests.add_test("test_color_management_effects_read", [] {
        using namespace otio;

        otio::ErrorStatus              status;
        SerializableObject::Retainer<> so =
            SerializableObject::from_json_string(
                R"(
            {
                "OTIO_SCHEMA": "Clip.1",
                "media_reference": {
                    "OTIO_SCHEMA": "ExternalReference.1",
                    "target_url": "unit_test_url"
                },
                "effects": [
                    {
                        "OTIO_SCHEMA": "VideoBrightness.1",
                        "name": "brightness",
                        "brightness": 50,
                        "effect_name": "VideoBrightness",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoContrast.1",
                        "name": "contrast",
                        "contrast": 20,
                        "effect_name": "VideoContrast",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoSaturation.1",
                        "name": "saturation",
                        "saturation": 70,
                        "effect_name": "VideoSaturation",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoLightness.1",
                        "name": "lightness",
                        "lightness": 10,
                        "effect_name": "VideoLightness",
                        "enabled": true
                    },
                    {
                        "OTIO_SCHEMA": "VideoColorTemperature.1",
                        "name": "temperature",
                        "temperature": 6500,
                        "effect_name": "VideoColorTemperature",
                        "enabled": true
                    }
                ]
            })",
                &status);

        if (is_error(status))
            throw std::invalid_argument(status.details);

        const Clip* clip = dynamic_cast<const Clip*>(so.value);
        assertNotNull(clip);

        auto effects = clip->effects();
        assertEqual(effects.size(), 5);

        auto video_brightness = dynamic_cast<const VideoBrightness*>(effects[0].value);
        assertNotNull(video_brightness);
        assertEqual(video_brightness->brightness(), 50);

        auto video_contrast = dynamic_cast<const VideoContrast*>(effects[1].value);
        assertNotNull(video_contrast);
        assertEqual(video_contrast->contrast(), 20);

        auto video_saturation = dynamic_cast<const VideoSaturation*>(effects[2].value);
        assertNotNull(video_saturation);
        assertEqual(video_saturation->saturation(), 70);

        auto video_lightness = dynamic_cast<const VideoLightness*>(effects[3].value);
        assertNotNull(video_lightness);
        assertEqual(video_lightness->lightness(), 10);

        auto video_temperature = dynamic_cast<const VideoColorTemperature*>(effects[4].value);
        assertNotNull(video_temperature);
        assertEqual(video_temperature->temperature(), 6500);
    });

    tests.add_test("test_color_management_effects_write", [] {
        using namespace otio;

        SerializableObject::Retainer<otio::Clip> clip(new otio::Clip(
            "unit_clip",
            new otio::ExternalReference("unit_test_url"),
            std::nullopt,
            otio::AnyDictionary(),
            { new otio::VideoBrightness("brightness", 50),
              new otio::VideoContrast("contrast", 20),
              new otio::VideoSaturation("saturation", 70),
              new otio::VideoLightness("lightness", 10),
              new otio::VideoColorTemperature("temperature", 6500)}));

        auto json = clip.value->to_json_string();

        std::string expected_json = R"({
    "OTIO_SCHEMA": "Clip.2",
    "metadata": {},
    "name": "unit_clip",
    "source_range": null,
    "effects": [
        {
            "OTIO_SCHEMA": "VideoBrightness.1",
            "metadata": {},
            "name": "brightness",
            "effect_name": "VideoBrightness",
            "enabled": true,
            "brightness": 50
        },
        {
            "OTIO_SCHEMA": "VideoContrast.1",
            "metadata": {},
            "name": "contrast",
            "effect_name": "VideoContrast",
            "enabled": true,
            "contrast": 20
        },
        {
            "OTIO_SCHEMA": "VideoSaturation.1",
            "metadata": {},
            "name": "saturation",
            "effect_name": "VideoSaturation",
            "enabled": true,
            "saturation": 70
        },
        {
            "OTIO_SCHEMA": "VideoLightness.1",
            "metadata": {},
            "name": "lightness",
            "effect_name": "VideoLightness",
            "enabled": true,
            "lightness": 10
        },
        {
            "OTIO_SCHEMA": "VideoColorTemperature.1",
            "metadata": {},
            "name": "temperature",
            "effect_name": "VideoColorTemperature",
            "enabled": true,
            "temperature": 6500
        }
    ],
    "markers": [],
    "enabled": true,
    "media_references": {
        "DEFAULT_MEDIA": {
            "OTIO_SCHEMA": "ExternalReference.1",
            "metadata": {},
            "name": "",
            "available_range": null,
            "available_image_bounds": null,
            "target_url": "unit_test_url"
        }
    },
    "active_media_reference_key": "DEFAULT_MEDIA"
})";

        assertEqual(json, expected_json);
    });

    tests.run(argc, argv);
    return 0;
}
